contracts/distribution/dao-rewards-distributor/src/contract.rs

```diff
@@ -279,36 +279,106 @@ fn execute_fund_native(
 }
 
 fn execute_fund(
+    deps: DepsMut,
+    env: Env,
+    distribution: DistributionState,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    match distribution.active_epoch.emission_rate {
+        EmissionRate::Paused {} => execute_fund_paused(deps, distribution, amount),
+        EmissionRate::Immediate {} => execute_fund_immediate(deps, env, distribution, amount),
+        EmissionRate::Linear { .. } => execute_fund_linear(deps, env, distribution, amount),
+    }
+}
+
+/// funding a paused distribution simply increases the funded amount.
+fn execute_fund_paused(
+    deps: DepsMut,
+    mut distribution: DistributionState,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    distribution.funded_amount += amount;
+
+    DISTRIBUTIONS.save(deps.storage, distribution.id, &distribution)?;
+
+    Ok(Response::new()
+        .add_attribute("action", "fund")
+        .add_attribute("id", distribution.id.to_string())
+        .add_attribute("denom", distribution.get_denom_string())
+        .add_attribute("amount_funded", amount))
+}
+
+/// funding an immediate distribution instantly distributes the new amount.
+fn execute_fund_immediate(
+    deps: DepsMut,
+    env: Env,
+    mut distribution: DistributionState,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    distribution.funded_amount += amount;
+
+    // for immediate distribution, update total_earned_puvp instantly since we
+    // need to know the change in funded_amount to calculate the new
+    // total_earned_puvp.
+    distribution.update_immediate_emission_total_earned_puvp(deps.as_ref(), &env.block, amount)?;
+
+    DISTRIBUTIONS.save(deps.storage, distribution.id, &distribution)?;
+
+    Ok(Response::new()
+        .add_attribute("action", "fund")
+        .add_attribute("id", distribution.id.to_string())
+        .add_attribute("denom", distribution.get_denom_string())
+        .add_attribute("amount_funded", amount))
+}
+
+/// funding a linear distribution requires some complex logic based on whether
+/// or not the distribution is continuous and whether or not it's expired.
+///
+/// expired continuous distributions experience backfill with the new funds,
+/// whereas expired discontinuous distributions begin anew (and all past rewards
+/// must be taken into account before restarting distribution).
+fn execute_fund_linear(
     deps: DepsMut,
     env: Env,
     mut distribution: DistributionState,
     amount: Uint128,
 ) -> Result<Response, ContractError> {
-    // will only be true if emission rate is linear and continuous is true
     let continuous =
         if let EmissionRate::Linear { continuous, .. } = distribution.active_epoch.emission_rate {
             continuous
         } else {
             false
         };
+    let previously_funded = !distribution.funded_amount.is_zero();
+    let was_expired = distribution.active_epoch.ends_at.is_expired(&env.block);
+    let discontinuous_expired = !continuous && was_expired;
```

Comment:
should we refer to previous expirations as discrete rather than discontinuous? latter kind of gives the impression that the rewards are discontinued and not to be restarted, but maybe that's just me 
```suggestion
    let discrete_expired = !continuous && was_expired;
```

Reply:
hmm, i think continuous has a different meaning in this context. a continuous distribution is one that retroactively distributes rewards when when you fund it after its rewards have run out. a discontinuous distribution starts distributing new funds from the current block if it has previously expired. discrete isn't a meaningful term here, though i'm open to better naming if you have other suggestions!